q2_types/per_sample_sequences/_format.py

```diff
@@ -105,25 +107,37 @@ class FastqGzFormat(model.BinaryFileFormat):
     A gzipped fastq file.
 
     """
-    def sniff(self):
+    def validate(self):
         with self.open() as fh:
-            if fh.read(2)[:2] != b'\x1f\x8b':
-                return False
-
-        filepath = str(self)
-        sniffer = skbio.io.io_registry.get_sniffer('fastq')
-        if sniffer(str(self))[0]:
-            try:
-                generator = skbio.io.read(filepath, constructor=skbio.DNA,
-                                          phred_offset=33, format='fastq',
-                                          verify=False)
-                for seq, _ in zip(generator, range(15)):
-                    pass
-                return True
-            # ValueError raised by skbio if there are invalid DNA chars.
-            except ValueError:
-                pass
-        return False
+            if fh.peek(2)[:2] != b'\x1f\x8b':
+                raise ValidationError('File is uncompressed')
+
+        with gzip.open(str(self), mode='rt', encoding='ascii') as fh:
+            for i, record in enumerate(itertools.zip_longest(*[fh] * 4)):
+                header, seq, sep, qual = record
+
+                if not header.startswith('@'):
+                    raise ValidationError('Header on line %d is not FASTQ, '
+                                          'records may be misaligned' %
+                                          (i * 4 + 1))
+
+                if seq is None:
+                    raise ValidationError('Missing sequence for record '
+                                          'beginning on line %d'
+                                          % (i * 4 + 1))
+                elif not seq.isupper():
```

Comment:
Since `.isupper()` takes a linear pass, I wonder if we could validate the sequence using `skbio.DNA(seq, validate=True)` instead? If the performance is comparable it'd probably be worth having the extra validation.

Reply:
DNA's constructor is pretty slow and Python str operations are super fast, our reasoning was that it's pretty unlikely that a user would have a fastq file with the wrong characters which didn't also cause the `@` check to throw up. And a protein fastq doesn't really make sense. Is it typical that the sequence's characters are the issue with fastq?

Reply:
Just tested:
```python
In [1]: import skbio

In [2]: a = "ACGT" * 50

In [3]: %timeit a.isupper()
The slowest run took 4.55 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.1 µs per loop

In [4]: %timeit skbio.DNA(a, validate=True)
The slowest run took 11.97 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 15.9 µs per loop
```

Reply:
Performance hit doesn't seem worth it. Thanks for profiling!